gps2val/gpsd_feeder.py

```diff
@@ -20,115 +20,276 @@
+    else:
+        formatter = logging.Formatter(
+            fmt="%(asctime)s %(levelname)s %(name)s: %(message)s"
+        )
+
+    # add formatter to console_logger
+    console_logger.setFormatter(formatter)
+
+    # add console_logger as a global handler
+    root_logger = logging.getLogger()
+    root_logger.setLevel(loglevel)
+    root_logger.addHandler(console_logger)
+
+
+class ColorFormatter(logging.Formatter):
+    FORMAT = "{time} {{loglevel}} {logger} {msg}".format(
+        time="\x1b[2m%(asctime)s\x1b[0m",  # grey
+        logger="\x1b[2m%(name)s:\x1b[0m",  # grey
+        msg="%(message)s",
+    )
+    FORMATS = {
+        logging.DEBUG: FORMAT.format(loglevel="\x1b[34mDEBUG\x1b[0m"),  # blue
+        logging.INFO: FORMAT.format(loglevel="\x1b[32mINFO\x1b[0m"),  # green
+        logging.WARNING: FORMAT.format(loglevel="\x1b[33mWARNING\x1b[0m"),  # yellow
+        logging.ERROR: FORMAT.format(loglevel="\x1b[31mERROR\x1b[0m"),  # red
+        logging.CRITICAL: FORMAT.format(loglevel="\x1b[31mCRITICAL\x1b[0m"),  # red
+    }
+
+    def format(self, record):
+        log_fmt = self.FORMATS.get(record.levelno)
+        formatter = logging.Formatter(log_fmt)
+        return formatter.format(record)
+
+
+def parse_env_log(env_log, default=logging.INFO):
+    def parse_level(level, default=default):
+        if type(level) is str:
+            if level.lower() in [
+                "debug",
+                "info",
+                "warn",
+                "warning",
+                "error",
+                "critical",
+            ]:
+                return level.upper()
+            else:
+                raise Exception(f"could not parse '{level}' as a log level")
+        return default
+
+    loglevels = {}
+
+    if env_log is not None:
+        log_specs = env_log.split(",")
+        for log_spec in log_specs:
+            spec_parts = log_spec.split("=")
+            if len(spec_parts) == 1:
+                # This is a root level spec
+                if "root" in loglevels:
+                    raise Exception("multiple root loglevels specified")
+                else:
+                    loglevels["root"] = parse_level(spec_parts[0])
+            if len(spec_parts) == 2:
+                logger = spec_parts[0]
+                level = spec_parts[1]
+                loglevels[logger] = parse_level(level)
+
+    if "root" not in loglevels:
+        loglevels["root"] = default
+
+    return loglevels
+
 
 class Kuksa_Client():
 
     # Constructor
     def __init__(self, config):
-        print("Init kuksa client...")
+        log.info("Init kuksa client...")
         if "kuksa_val" not in config:
-            print("kuksa_val section missing from configuration, exiting")
+            log.error("kuksa_val section missing from configuration, exiting")
             sys.exit(-1)
-        provider_config=config['kuksa_val']
+        provider_config = config['kuksa_val']
         self.client = KuksaClientThread(provider_config)
```

Comment:
If you use this threaded API, I think the lines 133 to 135 are unnecessary. What this does is ensure a startup connection. Therfore it takes the token specified in the config and tries to call a ```GetServerInfo()``` If this fails it will fail too and not startup.

Reply:
Also dependant on the fix https://github.com/eclipse/kuksa.val/pull/530

Reply:
As of today, if using latest master from [kuksa](https://github.com/eclipse/kuksa.val), the `authorize` on line 135 is needed, if you remove it you will get errors that you are not authenticated, for both KUKSA.val Server and Databroker.

Broker with line 135:

```
erik@debian3:~/kuksa.val.feeders/gps2val$ python gpsd_feeder.py --protocol grpc --port 55555 --insecure true --token /home/erik/kuksa.val/jwt/provide-all.token 
2023-04-19 14:52:51,798 INFO gpsfeeder: No configuration file found. Using default values.
2023-04-19 14:52:51,799 INFO gpsfeeder: Init kuksa client...
2023-04-19 14:52:51,838 INFO gpsfeeder: Authorizing using /home/erik/kuksa.val/jwt/provide-all.token
gRPC channel connected.
2023-04-19 14:52:51,847 INFO gpsfeeder: Connection status is: True
2023-04-19 14:52:51,847 INFO gpsfeeder: Init gpsd client...
2023-04-19 14:52:51,847 INFO gpsfeeder: Trying to connect gpsd at 127.0.0.1 port 2948
2023-04-19 14:52:51,847 INFO gpsfeeder: Using mapping
2023-04-19 14:52:51,848 INFO gpsfeeder: gpsd receive loop started
2023-04-19 14:52:52,667 INFO gpsfeeder: Number of VSS messages sent so far: 1
2023-04-19 14:52:52,686 INFO gpsfeeder: Number of VSS messages sent so far: 2
2023-04-19 14:52:52,728 INFO gpsfeeder: Number of VSS messages sent so far: 4
2023-04-19 14:52:53,831 INFO gpsfeeder: Number of VSS messages sent so far: 8
2023-04-19 14:52:54,981 INFO gpsfeeder: Number of VSS messages sent so far: 16
2023-04-19 14:52:57,335 INFO gpsfeeder: Number of VSS messages sent so far: 32
```

Broker with line 135 commented:

```
erik@debian3:~/kuksa.val.feeders/gps2val$ python gpsd_feeder.py --protocol grpc --port 55555 --insecure true --token /home/erik/kuksa.val/jwt/provide-all.token 
2023-04-19 14:54:49,313 INFO gpsfeeder: No configuration file found. Using default values.
2023-04-19 14:54:49,314 INFO gpsfeeder: Init kuksa client...
2023-04-19 14:54:49,356 INFO gpsfeeder: Authorizing using /home/erik/kuksa.val/jwt/provide-all.token
2023-04-19 14:54:49,356 INFO gpsfeeder: Connection status is: False
2023-04-19 14:54:49,356 INFO gpsfeeder: Init gpsd client...
2023-04-19 14:54:49,356 INFO gpsfeeder: Trying to connect gpsd at 127.0.0.1 port 2948
2023-04-19 14:54:49,357 INFO gpsfeeder: Using mapping
2023-04-19 14:54:49,359 INFO gpsfeeder: gpsd receive loop started
gRPC channel connected.
2023-04-19 14:54:50,202 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:54:50,214 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:54:50,229 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:54:50,241 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:54:50,252 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:54:50,264 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:54:50,276 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
```
Server without 135 commented:

```
erik@debian3:~/kuksa.val.feeders/gps2val$ python gpsd_feeder.py --file config.ini
2023-04-19 14:57:19,706 INFO gpsfeeder: # Using config from: config.ini
2023-04-19 14:57:19,706 INFO gpsfeeder: Init kuksa client...
2023-04-19 14:57:19,804 INFO gpsfeeder: Authorizing using /home/erik/kuksa.val/jwt/provide-all.token
gRPC channel connected.
2023-04-19 14:57:19,819 INFO gpsfeeder: Connection status is: True
2023-04-19 14:57:19,819 INFO gpsfeeder: Init gpsd client...
2023-04-19 14:57:19,819 INFO gpsfeeder: Trying to connect gpsd at 127.0.0.1 port 2948
2023-04-19 14:57:19,819 INFO gpsfeeder: Using mapping
2023-04-19 14:57:19,824 INFO gpsfeeder: gpsd receive loop started
2023-04-19 14:57:23,243 INFO gpsfeeder: Number of VSS messages sent so far: 1
2023-04-19 14:57:23,276 INFO gpsfeeder: Number of VSS messages sent so far: 2
2023-04-19 14:57:23,326 INFO gpsfeeder: Number of VSS messages sent so far: 4
2023-04-19 14:57:23,448 INFO gpsfeeder: Number of VSS messages sent so far: 8
```

Server with 135 commented:

```
erik@debian3:~/kuksa.val.feeders/gps2val$ !5701
python gpsd_feeder.py --file config.ini
2023-04-19 14:55:59,697 INFO gpsfeeder: # Using config from: config.ini
2023-04-19 14:55:59,697 INFO gpsfeeder: Init kuksa client...
2023-04-19 14:55:59,753 INFO gpsfeeder: Authorizing using /home/erik/kuksa.val/jwt/provide-all.token
2023-04-19 14:55:59,753 INFO gpsfeeder: Connection status is: False
2023-04-19 14:55:59,754 INFO gpsfeeder: Init gpsd client...
2023-04-19 14:55:59,754 INFO gpsfeeder: Trying to connect gpsd at 127.0.0.1 port 2948
2023-04-19 14:55:59,754 INFO gpsfeeder: Using mapping
gRPC channel connected.
2023-04-19 14:55:59,759 INFO gpsfeeder: gpsd receive loop started
2023-04-19 14:56:19,710 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:19,724 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:19,737 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:19,754 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:19,766 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:19,785 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:20,800 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:20,814 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:20,829 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:20,842 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:20,857 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:20,870 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:20,883 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:20,895 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:21,906 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:21,920 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:21,934 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:21,948 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:21,961 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:21,974 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}
2023-04-19 14:56:21,987 ERROR gpsfeeder: Error sending signal: {'code': 16, 'reason': 'unauthenticated', 'message': 'Invalid auth token'}

```

Reply:
After https://github.com/eclipse/kuksa.val/pull/530 was merged the line 135 does not any longer seem to be necessary. I have updated the PR